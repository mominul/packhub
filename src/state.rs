use std::sync::Arc;

use anyhow::Result;
use dotenvy::var;
use mongodb::Client;
use sequoia_openpgp::{serialize::SerializeInto, Cert};

use crate::pgp::{
    clearsign_metadata, detached_sign_metadata, generate_and_save_keys, load_cert_from_file,
};

#[derive(Clone)]
pub struct AppState {
    state: Arc<InnerState>,
}

struct InnerState {
    db: Client,
    cert: Cert,
}

impl AppState {
    pub async fn initialize(generate_keys: bool) -> Self {
        let uri = format!(
            "mongodb://{}:{}@{}:27017",
            var("PACKHUB_DB_USER").unwrap(),
            var("PACKHUB_DB_PASSWORD").unwrap(),
            var("PACKHUB_DB_HOST").unwrap()
        );

        let client = Client::with_uri_str(uri).await.unwrap();

        let cert = if generate_keys {
            generate_and_save_keys().unwrap()
        } else {
            load_cert_from_file().unwrap()
        };

        Self {
            state: Arc::new(InnerState { db: client, cert }),
        }
    }

    /// Get a reference to the MongoDB client.
    pub fn db(&self) -> &Client {
        &self.state.db
    }

    pub fn clearsign_metadata(&self, data: &str) -> Result<String> {
        clearsign_metadata(data, &self.state.cert)
    }

    pub fn detached_sign_metadata(&self, data: &str) -> Result<String> {
        detached_sign_metadata(data, &self.state.cert)
    }

    pub fn armored_public_key(&self) -> Vec<u8> {
        self.state.cert.armored().to_vec().unwrap()
    }

    pub fn dearmored_public_key(&self) -> Vec<u8> {
        self.state.cert.to_vec().unwrap()
    }
}
