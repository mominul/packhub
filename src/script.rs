use askama::Template;
use axum::{extract::Path, routing::get, Router};

use crate::state::AppState;

#[derive(Template)]
#[template(path = "apt-script.sh", escape = "none")]
struct AptScript<'a> {
    host: &'a str,
    distro: &'a str,
    owner: &'a str,
    repo: &'a str,
}

fn generate_apt_script(distro: &str, owner: &str, repo: &str) -> String {
    let host = dotenvy::var("PACKHUB_DOMAIN").unwrap();
    let script = AptScript {
        host: &host,
        distro,
        owner,
        repo,
    };
    script.render().unwrap()
}

async fn apt_script_handler(Path((distro, owner, repo)): Path<(String, String, String)>) -> String {
    generate_apt_script(&distro, &owner, &repo)
}

pub fn script_routes() -> Router<AppState> {
    Router::new().route("/{distro}/github/{owner}/{repo}", get(apt_script_handler))
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_script_generation_apt() {
        let apt_script = generate_apt_script("ubuntu", "OpenBangla", "OpenBangla-Keyboard");
        assert_snapshot!(apt_script);
    }
}
