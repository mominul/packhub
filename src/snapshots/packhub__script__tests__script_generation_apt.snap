---
source: src/script.rs
expression: apt_script
---
#!/bin/sh

echo "Welcome to package key and repository setup script for OpenBangla-Keyboard"
echo "This script will add the repository key and repository to your system."
echo "Please make sure you have sudo access to run this script."
echo
echo "Downloading and installing the repository key..."
wget -qO- http://localhost:3000/v1/keys/packhub.gpg | sudo tee /etc/apt/keyrings/packhub.gpg > /dev/null
echo
echo "Adding the repository to your system..."
echo "deb [signed-by=/etc/apt/keyrings/packhub.gpg] http://localhost:3000/v1/apt/ubuntu/github/OpenBangla/OpenBangla-Keyboard stable main" | sudo tee /etc/apt/sources.list.d/OpenBangla-Keyboard.list > /dev/null
echo 
echo "Updating package lists..."
sudo apt-get update
